pkg/syncer/synctarget/shard_manager.go

```diff
@@ -0,0 +1,159 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package synctarget
+
+import (
+	"context"
+	"sync"
+
+	kcpcache "github.com/kcp-dev/apimachinery/v2/pkg/cache"
+	kcpdynamic "github.com/kcp-dev/client-go/dynamic"
+	"github.com/kcp-dev/logicalcluster/v3"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/klog/v2"
+
+	kcpcorev1alpha1 "github.com/kcp-dev/kcp/pkg/apis/core/v1alpha1"
+	workloadv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/workload/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/informer"
+)
+
+var logicalClusterGVR schema.GroupVersionResource = kcpcorev1alpha1.SchemeGroupVersion.WithResource("logicalclusters")
+
+// ShardAccess contains clustered dynamic clients, as well as
+// cluster-aware informer factories for both the Syncer and Upsyncer virtual workspaces
+// associated to a Shard.
+type ShardAccess struct {
+	SyncerClient   kcpdynamic.ClusterInterface
+	SyncerDDSIF    *informer.DiscoveringDynamicSharedInformerFactory
+	UpsyncerClient kcpdynamic.ClusterInterface
+	UpsyncerDDSIF  *informer.DiscoveringDynamicSharedInformerFactory
```

Comment:
what are these? Upstream/downstream? Why a new concept?

Reply:
There are shards only for upstream (== KCP)

Reply:
Just gathering the shard-related upstream clients and informers for both the Syncer and UpSyncer.

Naming suggestion welcomed :-)